src/NuGet.Core/NuGet.Protocol/Utility/ProtocolDiagnosticsStream.cs

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+
+namespace NuGet.Protocol.Utility
+{
+    internal class ProtocolDiagnosticsStream : Stream
+    {
+        private Stream _baseStream;
+        private ProtocolDiagnosticEvent _protocolDiagnosticEvent;
+        private Stopwatch _stopwatch;
+        private long _bytes;
+
+        internal ProtocolDiagnosticsStream(Stream baseStream, ProtocolDiagnosticEvent protocolDiagnosticEvent, Stopwatch stopwatch)
+        {
+            _baseStream = baseStream;
+            _protocolDiagnosticEvent = protocolDiagnosticEvent;
+            _stopwatch = stopwatch;
+            _bytes = 0;
+        }
+
+        public override bool CanRead => _baseStream.CanRead;
+
+        public override bool CanSeek => _baseStream.CanSeek;
+
+        public override bool CanWrite => _baseStream.CanWrite;
+
+        public override long Length => _baseStream.Length;
+
+        public override long Position { get => _baseStream.Position; set => _baseStream.Position = value; }
+
+        public override void Flush()
+        {
+            _baseStream.Flush();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
```

Comment:
Do you need to implement the async stream methods?

Reply:
The async read eventually calls into the sync read. When I tried using both, every download was double counted.

Reply:
Okay, sounds good.